crates/accelerate/src/quantum_circuit/py_ext.rs

```diff
@@ -0,0 +1,34 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+use pyo3::ffi::Py_ssize_t;
+use pyo3::prelude::*;
+use pyo3::types::{PyList, PyTuple};
+use pyo3::{ffi, AsPyPointer, PyNativeType};
+
+pub fn tuple_new(py: Python<'_>, items: Vec<PyObject>) -> Py<PyTuple> {
+    unsafe {
+        let ptr = ffi::PyTuple_New(items.len() as Py_ssize_t);
+        let tup: Py<PyTuple> = Py::from_owned_ptr(py, ptr);
+        for (i, obj) in items.into_iter().enumerate() {
+            ffi::PyTuple_SetItem(ptr, i as Py_ssize_t, obj.into_ptr());
+        }
+        tup
+    }
+}
+
+pub fn tuple_new_empty(py: Python<'_>) -> Py<PyTuple> {
+    unsafe { Py::from_owned_ptr(py, ffi::PyTuple_New(0)) }
+}
+
+pub fn tuple_from_list(list: &PyList) -> Py<PyTuple> {
+    unsafe { Py::from_owned_ptr(list.py(), ffi::PyList_AsTuple(list.as_ptr())) }
+}
```

Comment:
The reason for these is to skip the intermediate creation of owned references (`&PyAny`) when we're planning to directly store a newly created GIL-independent `Py` without dereferencing it. I'll add some comments to this file to explain this.

The PyO3 API methods you mention return owned references, which are internally registered with the current `GILPool`, and later deallocated when the pool goes out of scope. Since we don't dereference them, the tracking and deallocation effort is wasted time.

This is originally something I'd done while trying to work around what appears to be a performance issue in macOS's `libmalloc` (or maybe in the kernel's implementation of `madvise` that it calls out to). When the default system deallocator is used, deallocations suddenly become two orders of magnitude more expensive beyond a certain threshold of Rust objects in play, which we hit for large circuits. Starting the process with environment variable `MallocMediumZone=0` disables the problematic deallocator, but we don't do this by default. Even with it on, using these custom methods seems like an easy win since it's a very hot path.

On the topic of the macOS issues I'd uncovered, should we document anywhere in Qiskit that `MallocMediumZone=0` may be necessary when working with large circuits on macOS?

Reply:
Comment added to explain these in 0fda5372a76ba2a4163993fe7fcd03e44cd64582.

Reply:
It feels weird to me that this isn't easily possible directly from PyO3 safe code, but I guess that's the reason for PyO3/pyo3#3382, which I'm pretty excited for.